ksql-rest-app/src/test/java/io/confluent/ksql/rest/server/computation/StatementExecutorTest.java

```diff
@@ -412,14 +455,185 @@ public void shouldEnforceReferentialIntegrity() {
+    return mockQuery;
+  }
+
+  @Test
+  public void shouldSkipStartForTerminatedQueriesWhenReplayingLog() {
+    // Given:
+    final String csas = "CSAS";
+    final String terminate = "TERMINATE";
+    final QueryId queryId = new QueryId("csas-query-id");
+    final String name = "foo";
+    final PersistentQueryMetadata mockQuery = mockReplayCSAS(csas, name, queryId);
+    final Statement mockTerminate = mockTerminate(queryId);
+    expect(mockParser.parseSingleStatement(terminate)).andReturn(mockTerminate);
+    expect(mockEngine.terminateQuery(queryId, false)).andReturn(true);
+    expect(mockEngine.getPersistentQueries()).andReturn(Collections.emptySet());
+    replayAll();
+
+    // When:
+    statementExecutorWithMocks.handleRestoration(
+        ImmutableList.of(
+            new QueuedCommand(
+                new CommandId(Type.STREAM, name, Action.CREATE),
+                new Command(csas, Collections.emptyMap(), Collections.emptyMap())
+            ),
+            new QueuedCommand(
+                new CommandId(Type.TERMINATE, queryId.getId(), Action.EXECUTE),
+                new Command(terminate, Collections.emptyMap(), Collections.emptyMap())
+            )
+        )
+    );
+
+    // Then:
+    verify(mockParser, mockEngine, mockQuery);
+  }
 
+  @Test
+  public void shouldStartUnterminatedQueriesWhenReplayingLog() {
+    // Given:
+    final String csas = "CSAS ";
+    final List<String> streamNames = ImmutableList.of("foo", "bar", "baz");
+    final List<PersistentQueryMetadata> queries = streamNames.stream()
+        .map(name -> mockReplayCSAS(csas + name, name, new QueryId("query-" + name)))
+        .collect(Collectors.toList());
+    expect(mockEngine.getPersistentQueries()).andStubReturn(queries);
+    for (final PersistentQueryMetadata query : queries) {
+      query.start();
+      expectLastCall();
+    }
+    replayAll();
+
+    // When:
+    statementExecutorWithMocks.handleRestoration(
+        streamNames.stream()
+            .map(
+                name -> new QueuedCommand(
+                    new CommandId(Type.STREAM, name, Action.CREATE),
+                    new Command(csas + name, Collections.emptyMap(), Collections.emptyMap())
+                ))
+            .collect(Collectors.toList()))
+    ;
+
+    // Then:
+    queries.forEach(EasyMock::verify);
+    verify(mockParser, mockEngine);
+  }
+
+  @Test
+  public void shouldCascade4Dot1DropStreamCommand() {
+    // Given:
+    final String drop = "DROP";
+    final DropStream mockDropStream = mockDropStream("foo");
+    expect(mockParser.parseSingleStatement(drop)).andReturn(mockDropStream);
+    expect(mockEngine.getMetaStore()).andStubReturn(mockMetaStore);
+    expect(mockMetaStore.getSource("foo"))
+        .andStubReturn(mock(StructuredDataSource.class));
+    expect(mockMetaStore.getQueriesWithSink("foo"))
+        .andStubReturn(ImmutableSet.of("query-id"));
+    expect(mockEngine.terminateQuery(eq(new QueryId("query-id")), eq(false)))
```

Comment:
This is the bit we want this test to actually check, right? That this gets called.  But at the moment this gets lost in all the other mock setup. So if we can move the other stuff into methods it makes the test clearer.

Of course, once you merge to master and switch to mockito this becomes much clearer as you just need to use 

```
verify(mockEngine).terminateQuery(blah);
```